content/articles/2017-07-ggplot2-tidy-evaluation.Rmarkdown

```diff
@@ -0,0 +1,150 @@
+slug: ggplot2-tidy-evaluation
+author: Mara Averick
+categories: [package]
+description: >
+  Using tidy evaluation in ggplot2 3.0.0.
+photo:
+  url: https://unsplash.com/photos/8KfCR12oeUM
+  author: Christopher Burns
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(collapse = TRUE, comment = "#>", fig.retina = 2, dpi = 300)
+library(tidyverse)
+```
+<html>
+<style>
+h3 code {
+    font-size: 1em;
+    
+}
+</style>
+</html>
+## Tidy evaluation
+
+One of the biggest changes in [ggplot2](https://ggplot2.tidyverse.org/) 3.0.0 is support for [tidy evaluation](https://adv-r.hadley.nz/evaluation.html#tidy-evaluation), making it more programmable, and more consistent with the rest of the tidyverse. Though this introduces some breaking changes, we believe it to be worthwhile in the interest of improving future code. Here, we outline some of the more prominent of those changes. For our complete list of changes, including symptomatic error messages, please see the [Breaking changes](https://ggplot2.tidyverse.org/news/index.html#breaking-changes) section of the [release notes](https://ggplot2.tidyverse.org/news/index.html#ggplot2-3-0-0).
+
+### `aes()` contains quosures
+
+The primary developer-facing change is that `aes()` now contains [quosures](https://adv-r.hadley.nz/evaluation.html#quosures) (expression + environment pairs), rather than symbols. As a result, you'll need to take a different approach to extracting the information you need. Below, we use a "quote-and-unquote" pattern for the aesthetic parameter `x_var`.
+
+```{r aes-quo}
+x_var <- quo(cyl)
+y_var <- quo(mpg)
+
+by_cyl <- mtcars %>%
+  group_by(!!x_var) %>%
+  summarise(mean = mean(!!y_var))
+
+ggplot(by_cyl, aes(!!x_var, mean)) +
+  geom_point()
+```
+
+You can now use [quasiquotation](https://adv-r.hadley.nz/quasiquotation.html) in [`aes()`](https://ggplot2.tidyverse.org/reference/aes.html), [`facet_wrap()`](https://ggplot2.tidyverse.org/reference/facet_wrap.html), and [`facet_grid()`](https://ggplot2.tidyverse.org/reference/facet_grid.html). For `aes()`, quasiquotation (`!!`, `!!!`, `:=`) replaces [`aes_()`](https://ggplot2.tidyverse.org/reference/aes_.html) and [`aes_string()`](https://ggplot2.tidyverse.org/reference/aes_.html) (though these functions are being soft deprecated, and will be around for a while).
+
+Here we'll use quasiquotation in place of `aes_()` to make a function to make a pie chart that allows the user to avoid using quotation marks when they pass their arguments to the function.[^1] The initial function, below, would require that the user know the exact specifications to pass to `aes()`.
+
+```{r piechart-basic}
+piechart_basic <- function(data, mapping) {
+  ggplot(data, mapping) +
+    geom_bar(width = 1) +
+    coord_polar(theta = "y") +
+    xlab(NULL) +
+    ylab(NULL)
+}
+piechart_basic(mpg, aes(factor(1), fill = class))
+```
+
+By quoting and unquoting the arguments, the user can leave the arguments bare when calling the function.
```

Comment:
The key to calling a tidy evaluation function inside another function is to quote (with `enquo()`) and unquote (with `!!`):